Bugfix/prod hotfixes 20200318

j5awry opened the pull request:
Added prod hotfix, skipping unfound branches/repos. Parameterized parallel builds, allowings users to set number of jobs (with a default of 'as many as possible.'). Add in ways to install extras like tox and launchpadlib. Expand README with more information

j5awry committed:
Skip missing repositories

Instead of hard failing when a repository is not found, or other cloning
type errors, instead skip the repository and print a message regarding
what was skipped

j5awry committed:
parameterize parallel jobs and enhance readme

Parameterized parallel builds, allowings users to set
number of jobs (with a defaultof 'as many as possible.').
Add in ways to install extras like tox and
launchpadlib. Expand README with more information

rcj4747 commented:
You have a commit in the PR just called 'test' that adds a newline.  Please drop that commit.

rcj4747 reviewed (commented):
The title and description of the 'Bring in hotfixes' commit doesn't add value as-is and needs to be changed.

1. Don't mention 'CPC' on an external PR because the acronym doesn't have meaning outside the company (it hardly has meaning inside the company) ;)
2. These were 'hotfixes' but that doesn't tell us what they address/change just that someone was running with patches on the upstream code for some time.
3. Consider organizing the changes functionally by patch.

rcj4747 reviewed (changes requested):
The 'Add prod hotfixes, parameterize parallel' commit message should be renamed or reorganized.  "Add prod hotfixes" does not inform the reader as to what the changes do or why.

rcj4747 commented on README.md:
````diff
@@ -18,3 +19,63 @@ pip install -e .
+
+```
+launchpad:
+    owners:
+        # Launchpad username
+        LAUNCHPAD-USERNAME:
+            max-age: 30
+    branches:
+        # Full path to repo
+        LAUNCHPAD-BZR-REPO-PATH:
+            review-count: 2
````
Please describe the function of:

* 'owners' vs 'branches
* max-age
* review-count

rcj4747 replied:
@j5awry I would still like to see the field documented.

rcj4747 commented on README.md:
````diff
@@ -18,3 +19,63 @@ pip install -e .
+```
+NOTE: as of v. 0.2, `owners` and all options underneath are experimental 
+with no guarantee of working. If you'd like to improve this work, we'll happily
+take PRs!
````
'owners' was added 3 years ago (prior to v0.2) in commit 6edf6350.  What is the rationale for making this experimental? Was there some discussion of this policy?  I know that we've been using this in production for some time now.  Because I know that we've been using the feature forever the 2nd sentence reads as "if it's broken why don't you fix my code" and it rubs me the wrong way.

j5awry replied:
Had IRC conversation with @rcj4747 . Discussed the issues around the `owner` stanza and the best way of communicating the issue. At present `get_branches_for_owner` isn't working as expected. I'll be changing this to be clear that we are aware the capabilities aren't as we'd like, that it requires  more testing and development, and will add TODO items to the README

rcj4747 commented on README.md:
```diff
@@ -18,3 +19,63 @@ pip install -e .
+
+To use this, tox must be installed in the environment. 
+
+Furthermore, to use all the launchpad capabilities, launchpadlib must be installed.
```
Is "all the launchpad capabilities" or is "any launchpad capability" correct?  If it's any then I'd like to see launchpadlib by default as Canonical's code hosting solution is launchpad and it should work out of the box.  If it is only a subset of capabilities then the readme should inform the user what they are/aren't so they can choose.

rcj4747 replied:
Thinking about this more, is 'tox' and dependencies large enough to push off in 'extras' rather than just making it a normal requirement?  pip installing tox in a new virtualenv pulled in: toml, py, filelock, six, appdirs, distlib, virtualenv, pluggy, pyparsing, packaging, tox which totaled about 5mb

j5awry replied:
I think this might be a question for @fginther ? I wasn't sure if not adding tox and launchpadlib were oversights or if they were separated for a reason. I went with "separated for a reason." I actually agree that they should just get pulled in normally, but I'll defer to longer maintainers.

rcj4747 replied:
We chatted this morning and decided to include tox by default and a commit has been pushed for that.

j5awry commented:
rebased and pushed better commit messages.

rcj4747 commented on setup.py:
```diff
@@ -7,15 +7,20 @@
     from pip._internal.req import parse_requirements
 except ImportError: # for pip <= 9.0.3
     from pip.req import parse_requirements
+from copy import deepcopy
```
I don't see deepcopy used in this commit.

rcj4747 replied:
Fixed in latest commit.

j5awry committed:
Make tox and launchpadlib normal reqs

Make tox and launchpadlib standard requirements and drop the extras.
Change the README to better describe current known issues and added a
TODO. expanded tox section explanation.

j5awry committed:
Update configuration info in readme

Add more explicit definitions for configuration items in the readme

rcj4747 commented on README.md:
```diff
@@ -52,6 +52,29 @@ lp-git:
+
+#### Specific Configurations
+* `owners.$USERNAME.max-age`: Int. Max age of the MP to show in search results
+* `repos.$REPO-NAME.review-count`: Int. Number of reviews to show in the search results
```
Is this 'reviews' like reviewers on MPs or MPs themselves?

fginther replied:
FYI, this is meant to specify the minimum number of review boxes to display for an MP, implying a minimum number of reviewers.

rcj4747 reviewed (changes requested):
I've posted a question about the 'review-count' config option

fginther reviewed (commented):
Thanks for the much improved README.

j5awry committed:
Fix docs for repos.$REPO-NAME.review-count

rcj4747 reviewed (approved):
+1

philroche commented:
+1 Thanks
